do-place and do-route now produces artifacts even in the case of failure

oharboe opened the pull request:
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/discussions/1593

oharboe committed:
makefile: add check-place and check-route

Empty reports are generated if there are no errors.

This supports the Bazel use-case to create artifacts for inspection
in the case that detailed place, global place,
global route or detailed route fails.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/scripts/detail_place.tcl:
```diff
@@ -7,7 +7,15 @@ source $::env(PLATFORM_DIR)/setRC.tcl
+if {[info exists env(WRITE_ON_FAIL)] && $::env(WRITE_ON_FAIL)} {
+  set result [catch {detailed_placement} errMsg]
+  if {$result != 0} {
+    write_db $::env(RESULTS_DIR)/3_5_place_dp.odb
```
I disagree with the approach of writing the final odb file as if it is succeeded.  If you want to write 3_5_place_db-FAILED.odb that would make more sense to me.

oharboe replied:
That does not work and defeats the purpose of this PR: Bazel must have the same artifact file name, even if placement failed. Checking for valid placement is a separate step afterwards. This separates the concern of producing an artifact to be used or inspected and checking for failure.

maliberty replied:
If you write the success artifact how will make know that the step is actually failed?

oharboe replied:
implemented via place.ok and route.ok files.

oharboe commented on flow/Makefile:
```diff
@@ -651,10 +651,41 @@ $(eval $(call do-copy,3_place,3_5_place_dp.odb,))
+		exit 1 ; \
+	fi
+	$(UNSET_AND_MAKE) do-3_place do-3_place.sdc
+	echo >$(RESULTS_DIR)/place.ok 1
```
@maliberty This file contains success/failure

maliberty replied:
Make will not read the contents of the file on the next invocation and will think this step succeeded.

oharboe replied:
Indeed:

1. do-xxxx is not compatible w normal make target checking, you have to pick one or the other. This is fine.
2. I need to add check-place to do-cts and check-route to do-finish

oharboe committed:
makefile: break out UNSET_VARS and SUB_MAKE from UNSET_AND_MAKE

for subsequent work on do-place/route

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/Makefile:
```diff
@@ -657,10 +659,42 @@ $(eval $(call do-copy,3_place,3_5_place_dp.odb,))
+# placement did not complete successfully, because $(RESULTS_DIR)/place.ok
+# is set to 0.
+#
+# This means that build systems, such as Bazel, that require a non-zero
+# exit code and the same artifacts to be produced every time, or it
+# will not publish the artifacts, can use this target.
```
@QuantamHD how do your Bazel setup work?  Presumably you would have the same issue so I'm wondering if you solved it this way?

QuantamHD replied:
I break my flow into steps (placement, cts, drt, etc.). Run them one at a time, and if a step fails I just rerun the tcl script for that stage which references outputs from the previous successful stage.

Long story short if a step fails I don't get artifacts either.

oharboe replied:
I see.

My main use-cases for Bazel is that I want artifacts for projects that have hours of running time so that I can examine failed builds quickly. This allows me to turn around a tweak to my scripts in a short amount of time or run `make issue` so that I can easily send feature request/issues to the community(frequently using deltaDebug.py...)

This PR gives me artifacts for failed buidls for global place, detailed place, global route and detailed route.

Later, I suspect that I'll find it worth adding artifacts for failed builds to CTS.

Note! We are also using bazel_rules_hdl, but to create a Verilator artifact.

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

oharboe committed:
makefile: check place reuqires fix in global_route

$::env() is required in non-global context on some
tcl versions

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

oharboe committed:
makefile: do-floorplan creates 2_3_floorplan_tdms.log always

this way a Bazel layer on top does not need to know about
RTLMP_FLOW=True/False

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
makefile: move foo.tmp.log to foo.log in the case of failure

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
makefile: make sure that .log files exist for do-place and do-route

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
makefile: do-place, always create 5_2_fillcell.log

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
makefile: do-place/route now uses *-failed.odb to create artifact

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
makefile: remove SUB_MAKE and UNSET_VARS from make issue variables

fixes harmless errors

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

maliberty commented on flow/Makefile:
```diff
@@ -607,6 +609,7 @@ $(RESULTS_DIR)/2_floorplan.sdc: $(RESULTS_DIR)/2_1_floorplan.odb
 .PHONY: do-floorplan
 do-floorplan:
 	mkdir -p $(LOG_DIR) $(REPORTS_DIR)
+	touch $(LOG_DIR)/2_3_floorplan_tdms.log
```
Why?

maliberty commented on flow/Makefile:
```diff
@@ -658,10 +661,47 @@ $(eval $(call do-copy,3_place,3_5_place_dp.odb,))
+	@$(UNSET_VARS); \
+	$(SUB_MAKE) do-3_3_place_gp; \
+	if [ $$? -ne 0 ]; then \
+		cp $(RESULTS_DIR)/3_3_place_gp-failed.odb $(RESULTS_DIR)/3_place.odb ; \
```
This whole thing should be moved to Bazel not ORFS
